File: python/lmcc_std/tools.py
"""Tools and citations: formats and transports (spec/vocab/transport-tools.md,
transport-citations.md). Every value shape is lm15's; the program never
changes between the native and the text tier."""

from __future__ import annotations

import dataclasses

from lmcc.errors import refuse
from lmcc.formats import Format
from lmcc.transport import Transport

from . import jsontext
from .formats import lift

VERSION = "0.1.0"


# The host types a Python program spells; formats resolve by type, never
# by purpose (kernel §5), so ``tools`` and ``calls`` are different types.
# ``install`` binds them at runtime; an artifact names them under
# ``formats`` ("list[Tool]": {"use": "function_tool"}) to travel.

@dataclasses.dataclass
class Tool:
    name: str
    description: str | None = None
    parameters: dict | None = None


@dataclasses.dataclass
class ToolCall:
    id: str
    name: str
    input: dict


@dataclasses.dataclass
class Citation:
    url: str | None = None
    title: str | None = None
    text: str | None = None
    source: int | None = None


@dataclasses.dataclass
class Source:
    text: str
    title: str | None = None
    url: str | None = None

_TOOL_KEYS = {"name", "description", "parameters"}
_DEFAULT_PARAMETERS = {"type": "object", "properties": {}}


def _tool_items(value, field):
    items = value if isinstance(value, list) else [value]
    out = []
    for i, item in enumerate(items):
        spec = item if isinstance(item, dict) else {
            k: getattr(item, k) for k in _TOOL_KEYS if getattr(item, k, None) is not None}
        if not isinstance(spec.get("name"), str) or not spec["name"]:
            refuse("format-write-error", f"field {field.name!r}: tools[{i}] needs a string 'name'")
        unknown = set(spec) - _TOOL_KEYS - {"type"}
        if unknown:
            refuse("format-write-error",
                   f"field {field.name!r}: tools[{i}] has keys {sorted(unknown)}; a tool is "
                   f"name, description, parameters (lm15 FunctionTool)")
        out.append(spec)
    return out


class FunctionToolFormat(Format):
    """lm15 ``FunctionTool`` parts, for ``Request.tools``."""
    accepts = ("list[Tool]", "Tool", "list[*]", "object", "*")
    direction = "in"
    writes = "parts"
    reads = ("function",)

    def __init__(self, options: dict):
        pass

    def describe(self, field):
        return "tools"

    def write(self, value, field):
        parts = []
        for spec in _tool_items(value, field):
            part = {"type": "function", "name": spec["name"]}
            if spec.get("description"):
                part["description"] = spec["description"]
            part["parameters"] = spec.get("parameters") or dict(_DEFAULT_PARAMETERS)
            parts.append(part)
        return parts

    def read(self, capture, field):
        return [{k: v for k, v in p.items() if k != "type"} for p in capture.of("function")]


class ToolCatalogFormat(Format):
    """The same tools as text, for a model without native calling."""
    accepts = ("list[Tool]", "Tool", "list[*]", "object", "*")
    direction = "in"
    writes = "text"

    def __init__(self, options: dict):
        pass

    def describe(self, field):
        return "tools"

    def write(self, value, field):
        lines = []
        for spec in _tool_items(value, field):
            params = jsontext.dumps(spec.get("parameters") or _DEFAULT_PARAMETERS, indent=None)
            desc = f": {spec['description']}" if spec.get("description") else ""
            lines.append(f"- {spec['name']}({params}){desc}")
        return "\n".join(lines)


class ToolCallsFormat(Format):
    """Calls read from ``tool_call`` parts (id, name, input verbatim) or
    from text (fenced JSON; ids assigned ``call_1``… in reply order)."""
    accepts = ("list[ToolCall]", "list[*]", "*")
    direction = "both"
    writes = "parts"
    reads = ("tool_call", "text")

    def __init__(self, options: dict):
        pass

    def describe(self, field):
        return "tool calls"

    def write(self, value, field):
        out = []
        for c in value or []:
            c = dataclasses.asdict(c) if dataclasses.is_dataclass(c) else c
            out.append({"type": "tool_call", "id": c["id"], "name": c["name"], "input": c.get("input", {})})
        return out

    def read(self, capture, field):
        calls = []
        n = 0
        for p in capture.parts:
            if p.get("type") == "tool_call":
                calls.append({k: v for k, v in p.items() if k not in ("type", "continuation")})
            elif isinstance(p.get("text"), str):
                try:
                    obj = jsontext.loads(p["text"])
                except Exception as exc:  # noqa: BLE001
                    refuse("format-read-error", f"field {field.name!r}: a fenced call is not JSON: {exc}")
                if not isinstance(obj, dict) or not isinstance(obj.get("name"), str):
                    refuse("format-read-error", f"field {field.name!r}: a fenced call is {{name, input}}")
                n += 1
                calls.append({"id": f"call_{n}", "name": obj["name"], "input": obj.get("input") or {}})
        return lift(field.annotation, calls)


class CitationsFormat(Format):
    """From ``citation`` parts: url/title/text verbatim. From text markers:
    ``{"source": n}`` per distinct bracketed integer, order kept."""
    accepts = ("list[Citation]", "list[*]", "*")
    direction = "out"
    writes = "parts"
    reads = ("citation", "text")

    def __init__(self, options: dict):
        pass

    def read(self, capture, field):
        out, seen = [], set()
        for p in capture.parts:
            if p.get("type") == "citation":
                out.append({k: v for k, v in p.items() if k not in ("type", "continuation")})
            elif isinstance(p.get("text"), str):
                t = p["text"].strip()
                if t.isdigit() and t not in seen:
                    seen.add(t)
                    out.append({"source": int(t)})
        return lift(field.annotation, out)


class SourceListFormat(Format):
    """Numbered sources as text: ``[n] title: text``."""
    accepts = ("list[Source]", "list[*]", "*")
    direction = "in"
    writes = "text"

    def __init__(self, options: dict):
        pass

    def describe(self, field):
        return "numbered sources"

    def write(self, value, field):
        lines = []
        for i, s in enumerate(value or [], 1):
            s = dataclasses.asdict(s) if dataclasses.is_dataclass(s) else s
            if not isinstance(s, dict) or not isinstance(s.get("text"), str):
                refuse("format-write-error", f"field {field.name!r}: sources[{i - 1}] needs 'text'")
            title = s.get("title") or s.get("url") or f"source {i}"
            lines.append(f"[{i}] {title}: {s['text']}")
        return "\n".join(lines)


# ------------------------------------------------------------ transports

def native_tools(options: dict) -> Transport:
    return Transport(
        requires=["native_function_calling"], in_template=False,
        put={"@purpose": "request.tools"},
        find=[{"from": "part:tool_call", "to": "@purpose.calls", "complete_reply": True}])


FENCE_OPEN, FENCE_CLOSE = "```tool\n", "\n```"


def fenced_tools(options: dict) -> Transport:
    return Transport(
        requires=["instruct"], in_template=False,
        put={"@purpose": "message:system"}, written_as={"@purpose": "tool_catalog"},
        tell={"system": "You may call a tool by replying with exactly one fenced block:\n"
                             "```tool\n{\"name\": \"<tool>\", \"input\": {...}}\n```\n"
                             "and nothing else; you will be given the result and asked again."},
        find=[{"from": "text", "between": [FENCE_OPEN, FENCE_CLOSE], "to": "@purpose.calls",
                   "remove": True, "complete_reply": True}],
        spelling={"call": "```tool\n{\"name\": \"{name}\", \"input\": {input}}\n```",
               "result": "Result of {name} ({id}):\n{output}"})


def native_citations(options: dict) -> Transport:
    s = Transport(requires=["native_citations"], in_template=False,
                 find=[{"from": "part:citation", "to": "@purpose"}])
    if options.get("search", True):
        s.request_settings = {"tools": [{"type": "builtin", "name": "web_search"}]}
    return s


def inline_citations(options: dict) -> Transport:
    return Transport(
        requires=["instruct"], in_template=False,
        put={"@purpose.sources": "message:user"},
        tell={"system": "Cite the numbered sources inline as [n] after each claim they support."},
        find=[{"from": "text", "between": ["[", "]"], "to": "@purpose", "remove": False}])


def install(registry, *, exist_ok: bool = True) -> None:
    for name, fmt in (("function_tool", FunctionToolFormat), ("tool_catalog", ToolCatalogFormat),
                      ("tool_calls", ToolCallsFormat), ("citations", CitationsFormat),
                      ("source_list", SourceListFormat)):
        registry.register_format(name, fmt, version=VERSION, exist_ok=exist_ok)
    for name, factory in (("native_tools", native_tools), ("fenced_tools", fenced_tools),
                          ("native_citations", native_citations),
                          ("inline_citations", inline_citations)):
        registry.register_transport(name, factory, version=VERSION, exist_ok=exist_ok)
    # runtime type bindings for Python programs (kernel §5 step 3); an
    # artifact that must travel names the same formats under its `formats`
    registry.format(list[Tool], use="function_tool")
    registry.format(list[ToolCall], use="tool_calls")
    registry.format(list[Citation], use="citations")
    registry.format(list[Source], use="source_list")

File: go/lmccstd/tools.go
package lmccstd

// Tools and citations: formats and strategies (spec/vocab/strategy-tools.md,
// strategy-citations.md). Every value shape is lm15's; the program never
// changes between the native and the text tier.

import (
	"fmt"
	"strconv"
	"strings"

	"lmcc/lmcc"
)

var defaultParameters = func() *lmcc.Object { return lmcc.Obj("type", "object", "properties", lmcc.NewObject()) }

func toolItems(value any, f *lmcc.Field) ([]*lmcc.Object, error) {
	var items []any
	switch v := value.(type) {
	case []any:
		items = v
	default:
		items = []any{value}
	}
	out := make([]*lmcc.Object, 0, len(items))
	for i, item := range items {
		spec, ok := item.(*lmcc.Object)
		if !ok {
			return nil, fmt.Errorf("field %q: tools[%d] is not an object", f.Name, i)
		}
		if name, ok := spec.Str("name"); !ok || name == "" {
			return nil, fmt.Errorf("field %q: tools[%d] needs a string 'name'", f.Name, i)
		}
		for _, k := range spec.Keys {
			if k != "name" && k != "description" && k != "parameters" && k != "type" {
				return nil, fmt.Errorf("field %q: tools[%d] has key %q; a tool is name, description, parameters (lm15 FunctionTool)", f.Name, i, k)
			}
		}
		out = append(out, spec)
	}
	return out, nil
}

// ---- function_tool: lm15 FunctionTool parts for Request.tools

type FunctionToolFormat struct{ base }

func newFunctionToolFormat(*lmcc.Object) (lmcc.Format, error) {
	return &FunctionToolFormat{base{"function_tool"}}, nil
}
func (*FunctionToolFormat) Accepts() []string {
	return []string{"list[Tool]", "Tool", "list[*]", "object", "*"}
}
func (*FunctionToolFormat) Direction() string           { return "in" }
func (*FunctionToolFormat) Emits() string               { return "parts" }
func (*FunctionToolFormat) Reads() []string             { return []string{"function"} }
func (*FunctionToolFormat) Describe(*lmcc.Field) string { return "tools" }
func (*FunctionToolFormat) Write(value any, f *lmcc.Field) (any, error) {
	specs, err := toolItems(value, f)
	if err != nil {
		return nil, err
	}
	parts := []any{}
	for _, spec := range specs {
		name, _ := spec.Str("name")
		part := lmcc.Obj("type", "function", "name", name)
		if d, ok := spec.Str("description"); ok && d != "" {
			part.Set("description", d)
		}
		if params := spec.Object("parameters"); params != nil {
			part.Set("parameters", params.Clone())
		} else {
			part.Set("parameters", defaultParameters())
		}
		parts = append(parts, part)
	}
	return parts, nil
}
func (*FunctionToolFormat) Read(span lmcc.Span, f *lmcc.Field) (any, error) {
	out := []any{}
	for _, p := range span.Of("function") {
		out = append(out, without(p.(*lmcc.Object), "type"))
	}
	return out, nil
}

// ---- tool_catalog: the same tools as text

type ToolCatalogFormat struct{ base }

func newToolCatalogFormat(*lmcc.Object) (lmcc.Format, error) {
	return &ToolCatalogFormat{base{"tool_catalog"}}, nil
}
func (*ToolCatalogFormat) Accepts() []string {
	return []string{"list[Tool]", "Tool", "list[*]", "object", "*"}
}
func (*ToolCatalogFormat) Direction() string           { return "in" }
func (*ToolCatalogFormat) Describe(*lmcc.Field) string { return "tools" }
func (*ToolCatalogFormat) Write(value any, f *lmcc.Field) (any, error) {
	specs, err := toolItems(value, f)
	if err != nil {
		return nil, err
	}
	lines := make([]string, 0, len(specs))
	for _, spec := range specs {
		name, _ := spec.Str("name")
		var params any = defaultParameters()
		if p := spec.Object("parameters"); p != nil {
			params = p
		}
		line := "- " + name + "(" + lmcc.MarshalJSON(params, -1) + ")"
		if d, ok := spec.Str("description"); ok && d != "" {
			line += ": " + d
		}
		lines = append(lines, line)
	}
	return strings.Join(lines, "\n"), nil
}
func (*ToolCatalogFormat) Read(lmcc.Span, *lmcc.Field) (any, error) {
	return nil, fmt.Errorf("tool_catalog is input-only")
}

// ---- tool_calls: from tool_call parts, or fenced JSON text (ids assigned)

type ToolCallsFormat struct{ base }

func newToolCallsFormat(*lmcc.Object) (lmcc.Format, error) {
	return &ToolCallsFormat{base{"tool_calls"}}, nil
}
func (*ToolCallsFormat) Accepts() []string           { return []string{"list[ToolCall]", "list[*]", "*"} }
func (*ToolCallsFormat) Emits() string               { return "parts" }
func (*ToolCallsFormat) Reads() []string             { return []string{"tool_call", "text"} }
func (*ToolCallsFormat) Describe(*lmcc.Field) string { return "tool calls" }
func (*ToolCallsFormat) Write(value any, f *lmcc.Field) (any, error) {
	list, _ := value.([]any)
	parts := []any{}
	for _, c := range list {
		co, ok := c.(*lmcc.Object)
		if !ok {
			return nil, fmt.Errorf("field %q: a call is {id, name, input}", f.Name)
		}
		id, _ := co.Str("id")
		name, _ := co.Str("name")
		input, _ := co.Get("input")
		if input == nil {
			input = lmcc.NewObject()
		}
		parts = append(parts, lmcc.Obj("type", "tool_call", "id", id, "name", name, "input", input))
	}
	return parts, nil
}
func (*ToolCallsFormat) Read(span lmcc.Span, f *lmcc.Field) (any, error) {
	calls := []any{}
	n := 0
	for _, raw := range span.Parts {
		p, _ := raw.(*lmcc.Object)
		if t, _ := p.Str("type"); t == "tool_call" {
			calls = append(calls, without(p, "type", "continuation"))
			continue
		}
		text, ok := p.Str("text")
		if !ok {
			continue
		}
		v, err := lmcc.ParseJSON(text)
		if err != nil {
			return nil, fmt.Errorf("field %q: a fenced call is not JSON: %v", f.Name, err)
		}
		obj, ok := v.(*lmcc.Object)
		name, hasName := "", false
		if ok {
			name, hasName = obj.Str("name")
		}
		if !ok || !hasName {
			return nil, fmt.Errorf("field %q: a fenced call is {name, input}", f.Name)
		}
		n++
		input, _ := obj.Get("input")
		if input == nil {
			input = lmcc.NewObject()
		}
		calls = append(calls, lmcc.Obj("id", "call_"+strconv.Itoa(n), "name", name, "input", input))
	}
	return calls, nil
}

// ---- citations: from citation parts, or bracketed integers

type CitationsFormat struct{ base }

func newCitationsFormat(*lmcc.Object) (lmcc.Format, error) {
	return &CitationsFormat{base{"citations"}}, nil
}
func (*CitationsFormat) Accepts() []string           { return []string{"list[Citation]", "list[*]", "*"} }
func (*CitationsFormat) Direction() string           { return "out" }
func (*CitationsFormat) Emits() string               { return "parts" }
func (*CitationsFormat) Reads() []string             { return []string{"citation", "text"} }
func (*CitationsFormat) Describe(*lmcc.Field) string { return "citations" }
func (*CitationsFormat) Write(any, *lmcc.Field) (any, error) {
	return nil, fmt.Errorf("citations is output-only")
}
func (*CitationsFormat) Read(span lmcc.Span, f *lmcc.Field) (any, error) {
	out := []any{}
	seen := map[string]bool{}
	for _, raw := range span.Parts {
		p, _ := raw.(*lmcc.Object)
		if t, _ := p.Str("type"); t == "citation" {
			out = append(out, without(p, "type", "continuation"))
			continue
		}
		if text, ok := p.Str("text"); ok {
			t := lmcc.Strip(text)
			if n, err := strconv.Atoi(t); err == nil && n >= 0 && !seen[t] && isDigits(t) {
				seen[t] = true
				out = append(out, lmcc.Obj("source", int64(n)))
			}
		}
	}
	return out, nil
}

func isDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

// ---- source_list: numbered sources as text

type SourceListFormat struct{ base }

func newSourceListFormat(*lmcc.Object) (lmcc.Format, error) {
	return &SourceListFormat{base{"source_list"}}, nil
}
func (*SourceListFormat) Accepts() []string           { return []string{"list[Source]", "list[*]", "*"} }
func (*SourceListFormat) Direction() string           { return "in" }
func (*SourceListFormat) Describe(*lmcc.Field) string { return "numbered sources" }
func (*SourceListFormat) Write(value any, f *lmcc.Field) (any, error) {
	list, _ := value.([]any)
	lines := make([]string, 0, len(list))
	for i, raw := range list {
		s, ok := raw.(*lmcc.Object)
		text, hasText := "", false
		if ok {
			text, hasText = s.Str("text")
		}
		if !ok || !hasText {
			return nil, fmt.Errorf("field %q: sources[%d] needs 'text'", f.Name, i)
		}
		title, ok := s.Str("title")
		if !ok || title == "" {
			if u, ok := s.Str("url"); ok && u != "" {
				title = u
			} else {
				title = "source " + strconv.Itoa(i+1)
			}
		}
		lines = append(lines, "["+strconv.Itoa(i+1)+"] "+title+": "+text)
	}
	return strings.Join(lines, "\n"), nil
}
func (*SourceListFormat) Read(lmcc.Span, *lmcc.Field) (any, error) {
	return nil, fmt.Errorf("source_list is input-only")
}

func without(o *lmcc.Object, keys ...string) *lmcc.Object {
	out := lmcc.NewObject()
	for _, k := range o.Keys {
		skip := false
		for _, x := range keys {
			if k == x {
				skip = true
			}
		}
		if !skip {
			v, _ := o.Get(k)
			out.Set(k, lmcc.DeepClone(v))
		}
	}
	return out
}

// ---- strategies

func nativeTools(*lmcc.Object) (*lmcc.Strategy, error) {
	s := lmcc.NewStrategy()
	s.Requires = []string{"native_function_calling"}
	s.Visible = false
	s.Placement = lmcc.Obj("@role", "controls.tools")
	s.Routings = []*lmcc.Object{lmcc.Obj("from", "channel:tool_call", "to", "@role.calls", "suffices", true)}
	return s, nil
}

const fenceOpen, fenceClose = "```tool\n", "\n```"

func fencedTools(*lmcc.Object) (*lmcc.Strategy, error) {
	s := lmcc.NewStrategy()
	s.Requires = []string{"instruct"}
	s.Visible = false
	s.Placement = lmcc.Obj("@role", "message:system")
	s.Via = lmcc.Obj("@role", "tool_catalog")
	s.Fragments = lmcc.Obj("system", "You may call a tool by replying with exactly one fenced block:\n```tool\n{\"name\": \"<tool>\", \"input\": {...}}\n```\nand nothing else; you will be given the result and asked again.")
	s.Routings = []*lmcc.Object{lmcc.Obj("from", "text", "between", []any{fenceOpen, fenceClose}, "to", "@role.calls", "consume", true, "suffices", true)}
	s.Turns = lmcc.Obj("call", "```tool\n{\"name\": \"{name}\", \"input\": {input}}\n```", "result", "Result of {name} ({id}):\n{output}")
	return s, nil
}

func nativeCitations(options *lmcc.Object) (*lmcc.Strategy, error) {
	s := lmcc.NewStrategy()
	s.Requires = []string{"native_citations"}
	s.Visible = false
	s.Routings = []*lmcc.Object{lmcc.Obj("from", "channel:citation", "to", "@role")}
	if options.Bool("search", true) {
		s.Controls = lmcc.Obj("tools", []any{lmcc.Obj("type", "builtin", "name", "web_search")})
	}
	return s, nil
}

func inlineCitations(*lmcc.Object) (*lmcc.Strategy, error) {
	s := lmcc.NewStrategy()
	s.Requires = []string{"instruct"}
	s.Visible = false
	s.Placement = lmcc.Obj("@role.sources", "message:user")
	s.Fragments = lmcc.Obj("system", "Cite the numbered sources inline as [n] after each claim they support.")
	s.Routings = []*lmcc.Object{lmcc.Obj("from", "text", "between", []any{"[", "]"}, "to", "@role", "consume", false)}
	return s, nil
}
